pkg/apis/istio/v1alpha1/istioinstaller_types.proto

```diff
@@ -0,0 +1,461 @@
+    repeated string loadBalancerSourceRanges = 9;
+
+    // externalName is the external reference that kubedns or equivalent will
+    // return as a CNAME record for this service. No proxying will be involved.
+    // Must be a valid RFC-1123 hostname (https://tools.ietf.org/html/rfc1123)
+    // and requires Type to be ExternalName.
+    // +optional
+    string externalName = 10;
+
+    // externalTrafficPolicy denotes if this Service desires to route external
+    // traffic to node-local or cluster-wide endpoints. "Local" preserves the
+    // client source IP and avoids a second hop for LoadBalancer and Nodeport
+    // type services, but risks potentially imbalanced traffic spreading.
+    // "Cluster" obscures the client source IP and may cause a second hop to
+    // another node, but should have good overall load-spreading.
+    // +optional
+    string externalTrafficPolicy = 11;
+}
+
+// OperatorConfig defines a config used to install an external operator.
+message OperatorSpec {
+    // manifest_path is the URL for the operator install manifest.
+    string manifest_path = 1;
+
+    // namespace is the namespace the operator manifest and CR are installed into.
+    string namespace = 2;
+
+    // spec is the initial CR for the operator.
+    k8sObjectOverlay spec = 100;
+}
+
+// Patch for an existing k8s resource.
+message k8sObjectOverlay {
+    message PathValue {
+        // Path of the form a.b:c.e.:f
+        // Where b:c is a list element selector of the form key:value and :f is a list selector of the form :value.
+        // All path intermediate nodes must exist.
+        string path = 1;
+        // Value to add, delete or replace.
+        // For add, the path should be a new leaf.
+        // For delete, value should be unset.
+        // For replace, path should reference an existing node.
+        // All values are strings but are converted into appropriate type based on schema.
+//        Value                interface{}           `protobuf:"bytes,4,opt,name=value,proto3" json:"value,omitempty"`
+
+    }
+    // Resource API version.
+    string api_version = 1;
+    // Resource kind.
+    string kind = 2;
+    // Name of resource.
+    // Namespace is always the component namespace.
+    string name = 3;
+
+    // List of patches to apply to resource.
+    repeated PathValue patches = 4;
+}
+
+// Observed state of IstioInstall.
+message InstallStatus {
+    enum Status {
+        NONE = 0;
+        UPDATING = 1;
+        HEALTHY = 2;
+        ERROR = 3;
+    }
+    message VersionStatus {
+        string version = 1;
+        Status status = 2;
+    }
+
+    VersionStatus traffic_management = 1;
+    VersionStatus policy_telemetry = 2;
+    VersionStatus security = 3;
+    VersionStatus config_management = 4;
+
+    repeated VersionStatus ingress_gateway = 5;
+    repeated VersionStatus egress_gateway = 6;
+}
+
+// Mirrors k8s.io.api.core.v1.ResourceRequirements for unmarshaling.
+message Resources {
```

Comment:
if it mirrors it, would it not be better to simply embed the upstream type directly?  shallower learning curve if they're the same.

Reply:
The problem is that k8s protos have their own unmarshaler which doesn't extend to just plugging in these types into it. It's copied here so that standard jsonpb unmarshaler works and k8s APIs are preserved.
I was able to link to k8s definitions in some cases but not all due to weirdness in unmarshaling.